### PR TITLE
fix(gatsby): try to automatically recover when parcel segfaults

### PR DESCRIPTION
**Pull request by pieh**
## Description

Seems like ARM macs parcel (at least version we use) cause quite a bit of problems and results in often seg faults - this tries to run parcel in child process so we can try to handle segfaults by clearing just parcel cache and retrying (few times)

## Related Issues

Possibly fixes https://github.com/gatsbyjs/gatsby/issues/38483

### CODE REVIEW

File: packages/gatsby/src/utils/parcel/__tests__/compile-gatsby-files.ts [hunk 1]
```diff
@@ -20,6 +20,29 @@ const dir = {
...
+      super(workerPath, options)
+    }
+  }
+
+  return {
+    ...gatsbyWorker,
+    WorkerPool: WorkerPoolThatCanUseTS,
+  }
+})
+
```
Comment (pieh) [inline packages/gatsby/src/utils/parcel/__tests__/compile-gatsby-files.ts:45]: this is just a trick to be able to "load" TS file that is used in tests - in actually built package, regular CJS will be loaded

File: packages/gatsby/src/utils/parcel/compile-gatsby-files.ts [hunk 2]
```diff
@@ -107,33 +130,56 @@ export async function compileGatsbyFiles(
...
-    const { bundleGraph } = await parcel.run()
-    let cacheClosePromise = Promise.resolve()
+    let bundles: RunParcelReturn
     try {
-      // @ts-ignore store is public field on LMDBCache class, but public interface for Cache
-      // doesn't have it. There doesn't seem to be proper public API for this, so we have to
-      // resort to reaching into internals. Just in case this is wrapped in try/catch if
-      // parcel changes internals in future (closing cache is only needed when retrying
-      // so the if the change happens we shouldn't fail on happy builds)
-      cacheClosePromise = cache.store.close()
```
Comment (pieh) [inline packages/gatsby/src/utils/parcel/compile-gatsby-files.ts:128]: we are killing child process so this should be done automatically on process exits

File: .jestSetup.js [hunk 3]
```diff
@@ -5,7 +5,30 @@ if (
...
+      }
+      super(workerPath, options)
+    }
+  }
+
+  return {
+    ...gatsbyWorker,
+    WorkerPool: WorkerPoolThatCanUseTS,
+  }
+})
```
Comment (pieh) [inline .jestSetup.js:34]: This is just a trick for unit tests to be able to use source TS files when using `gatsby-worker` / child process setup

In actual usage TS source files become CJS and those will be used (all e2e tests would be using that for example)

File: packages/gatsby/src/utils/parcel/__tests__/compile-gatsby-files.ts [hunk 4]
```diff
@@ -175,6 +176,37 @@ describe(`gatsby file compilation`, () => {
...
+      [MockFunction] {
+        "calls": Array [
+          Array [
+            Object {
+              "context": Object {
+                "filePath": "<PROJECT_ROOT>/gatsby-node.ts",
+                "generalMessage": "Expected ';', '}' or <eof>",
+                "hints": null,
+                "origin": "@parcel/transformer-js",
+                "specificMessage": "This is the expression part of an expression statement",
```
Comment (pieh) [inline packages/gatsby/src/utils/parcel/__tests__/compile-gatsby-files.ts:195]: there is improvements to be made here to point out specific location as error we get is like this:
```
{
      diagnostics: [
        {
          message: "Expected ';', '}' or <eof>",
          codeFrames: [
            {
              filePath: '/Users/misiek/dev/pgs-gatsby/packages/gatsby/src/utils/parcel/__tests__/fixtures/error-in-code-ts/gatsby-node.ts',
              codeHighlights: [
                {
                  message: 'This is the expression part of an expression statement',
                  start: { line: 5, column: 1 },
                  end: { line: 5, column: 4 }
                },
                {
                  message: null,
                  start: { line: 5, column: 6 },
                  end: { line: 5, column: 7 }
                }
              ]
            }
          ],
          hints: null,
          origin: '@parcel/transformer-js'
        }
      ]
    }
```
but it's out of scope of this PR - purpose here is to try to handle segfaults without introducing regression - this added tests is just just a test for case we didn't have (what's the behavior on ts files that are invalid) and that I did break before my most recent commit

**Comment by laurogripa** [comment]
Nice to know someone is working on this issue!

I'm trying to run your branch as a local version of Gatsby in my project because this issue makes it almost impossible for me to develop.
I added `"gatsby": "file:../gatsby"` to my package.json (the relative path is correct), the installation went okay but I'm getting this errors:

```
➜  erc55 git:(main) ✗ npm run develop
> erc55@0.2.15 develop
> gatsby develop
sh: gatsby: command not found

➜  erc55 git:(main) ✗ npx gatsby develop
npm ERR! Cannot read properties of undefined (reading 'replace')
```

Any suggestions on how to run your branch locally or should I just wait for it to get merged and published?

**Comment by pieh** [comment]
@laurogripa I published canary version with those changes `gatsby@5.14.0-alpha-fix-parcel-segfault.11` if you want to try it out

File: packages/gatsby/src/utils/parcel/__tests__/fixtures/error-in-code-ts/gatsby-node.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import { GatsbyNode } from "gatsby"
+import { working } from "../utils/say-what-ts"
+import { createPages } from "../utils/create-pages-ts"
+
+this is wrong syntax that should't compile
```
Comment (kathmbeck) [inline packages/gatsby/src/utils/parcel/__tests__/fixtures/error-in-code-ts/gatsby-node.ts:5]: Should this line be a comment?

Comment (kathmbeck) [reply packages/gatsby/src/utils/parcel/__tests__/fixtures/error-in-code-ts/gatsby-node.ts:5]: Oh or is this there to fail the compilation 👍

**Review by kathmbeck — APPROVED** [review]

File: packages/gatsby/src/utils/parcel/__tests__/fixtures/error-in-code-ts/gatsby-node.ts (hunk 5 above)
Comment (pieh) [reply packages/gatsby/src/utils/parcel/__tests__/fixtures/error-in-code-ts/gatsby-node.ts:5]: Yeah, that's test fixture to test behavior when there is error in source file that should result in compilation problem and test itself is here https://github.com/gatsbyjs/gatsby/blob/31e704918a82f3d79bc82c91bfa1e3cb248d3020/packages/gatsby/src/utils/parcel/__tests__/compile-gatsby-files.ts#L180-L210

I did break this behavior when iterating on this code and noticed we didn't have tests for this case so I added it as part of this PR

https://github.com/gatsbyjs/gatsby/pull/38773#discussion_r1433729682 this comment highlight opportunity to improve error message possibly, but this was just out of scope of the fix

**Pull request merged**